code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +164,246 @@ public void updateAccount(Account account, String name, String kind, String exte
+
+                Condition allMembers = DSL.falseCondition();
+                for (Member member : delete) {
+                    allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
+                            .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(member.getExternalIdType()))
+                            .and(PROJECT_MEMBER.PROJECT_ID.eq(project.getId())));
+                }
+                List<? extends ProjectMember> toDelete = create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch();
+                for (ProjectMember member : toDelete) {
+                    objectProcessManager.executeStandardProcess(StandardProcess.DEACTIVATE, member, null);
+                    objectProcessManager.executeStandardProcess(StandardProcess.REMOVE, member, null);
+                }
+                List<ProjectMember> newMembers = new ArrayList<>();
+                for (Member member : create) {
+                    newMembers.add(createProjectMember(project, member));
+                }
+                for (Member member : update) {
+                    newMembers.add(updateMemberRole(project, member));
+                }
+                newMembers.addAll(getProjectMembersByMembers(createIgnore, project.getId()));
+                return newMembers;
+            }
+        });
     }
 
-    @Inject
-    public void setGenericResourceDao(GenericResourceDao resourceDao) {
-        this.resourceDao = resourceDao;
+    private List<? extends ProjectMember> getProjectMembersByMembers(Set<Member> members, long projectId) {
+        Condition allMembers = DSL.falseCondition();
+        for (Member member: members){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(member.getExternalIdType()))
+                    .and(PROJECT_MEMBER.ROLE.in(member.getRole()))
+                    .and(PROJECT_MEMBER.PROJECT_ID.eq(projectId))
+                    .and(PROJECT_MEMBER.STATE.eq(CommonStatesConstants.ACTIVE))
+                    .and(PROJECT_MEMBER.REMOVED.isNull()));
+        }
+        return create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch();
+    }
+
+    private ProjectMember updateMemberRole(Account project, Member member) {
+        ProjectMemberRecord gotMember = create()
+                .selectFrom(PROJECT_MEMBER).where(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
+                .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(member.getExternalIdType()))
+                .and(PROJECT_MEMBER.PROJECT_ID.eq(project.getId()))
+                .and(PROJECT_MEMBER.REMOVED.isNull())).fetchOne();
+        Map<Object, Object> properties = new HashMap<>();
+        properties.put(PROJECT_MEMBER.ROLE, member.getRole());
+        objectProcessManager.executeStandardProcess(StandardProcess.UPDATE, gotMember, objectManager.convertToPropertiesFor(ProjectMember.class, properties));
+        return gotMember;
```

Comment:
really minor - we should probably call "reload" before returning the object so its returned with the updated properties